src/ceph-volume/ceph_volume/util/device.py

```diff
@@ -27,10 +27,13 @@ class Devices(object):
     """
 
     def __init__(self, devices=None):
+        lvs = lvm.Volumes()
         if not sys_info.devices:
-            sys_info.devices = disk.get_devices()
-        self.devices = [Device(k) for k in
-                            sys_info.devices.keys()]
+            sys_info.devices = disk.get_devices(lvs=lvs.dcopy())
+
+        self.devices = []
+        for k in sys_info.devices.keys():
+            self.devices.append(Device(k, lvs.dcopy()))
```

Comment:
Why the need for `dcopy` here, especially for every device? Wouldn't it be enough to simply pass the reference?

Maybe keeping `lvs` as a global namedtuple, similar to `sys_info` and `conf` (see `src/ceph-volume/ceph_volume/__init__.py`). A `get_lvs` function could take care of populating the lvs and returning the tuple, any function that adds or deletes lvs would have to update the namedtuple.

Reply:
> Why the need for dcopy here, especially for every device? Wouldn't it be enough to simply pass the reference?

Yes. Actually, after [this](https://github.com/ceph/ceph/pull/30102/files#diff-6993397b6f0532424603186ca62822bcR251) change a separate copy should never be required.

> Maybe keeping lvs as a global namedtuple, similar to sys_info and conf (see src/ceph-volume/ceph_volume/__init__.py). A get_lvs function could take care of populating the lvs and returning the tuple, any function that adds or deletes lvs would have to update the namedtuple.

I was thinking to do something same yesterday but I wasn't sure if that would be right, since, I am not well-acquainted with the codebase. But if you think the same, this is better.


What I was also thinking was that perhaps the helper methods like `is_lv()` and `get_lv()` should not create LVs list at all. Although creating list of LVs within these methods makes it very convenient, it's very easy to call them a bunch of times which in turn has heavy performance penalty.

Reply:
I'm good with this approach for now. I think its worth getting rid of those `dcopy` calls if they are unnecessary, but otherwise :+1: